src/librustc_mir/transform/qualify_consts.rs

```diff
@@ -0,0 +1,984 @@
+enum Mode {
+    Const,
+    Static,
+    StaticMut,
+    ConstFn,
+    Fn
+}
+
+impl fmt::Display for Mode {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            Mode::Const => write!(f, "constant"),
+            Mode::Static | Mode::StaticMut => write!(f, "static"),
+            Mode::ConstFn => write!(f, "constant function"),
+            Mode::Fn => write!(f, "function")
+        }
+    }
+}
+
+fn is_const_fn(tcx: &TyCtxt, def_id: DefId) -> bool {
+    if let Some(node_id) = tcx.map.as_local_node_id(def_id) {
+        let fn_like = FnLikeNode::from_node(tcx.map.get(node_id));
+        match fn_like.map(|f| f.kind()) {
+            Some(FnKind::ItemFn(_, _, _, c, _, _, _)) => {
+                c == hir::Constness::Const
+            }
+            Some(FnKind::Method(_, m, _, _)) => {
+                m.constness == hir::Constness::Const
+            }
+            _ => false
+        }
+    } else {
+        tcx.sess.cstore.is_const_fn(def_id)
+    }
+}
+
+struct Qualifier<'a, 'tcx: 'a> {
+    mode: Mode,
+    span: Span,
+    def_id: DefId,
+    mir: &'a Mir<'tcx>,
+    rpo: ReversePostorder<'a, 'tcx>,
+    tcx: &'a TyCtxt<'tcx>,
+    param_env: ty::ParameterEnvironment<'a, 'tcx>,
+    qualif_map: &'a mut DefIdMap<Qualif>,
+    mir_map: Option<&'a MirMap<'tcx>>,
+    temp_qualif: Vec<Option<Qualif>>,
+    return_qualif: Option<Qualif>,
+    qualif: Qualif,
+    const_fn_arg_vars: BitVector,
+    location: Location,
+    temp_promotion_state: Vec<TempState>,
+    promotion_candidates: Vec<Candidate>
+}
+
+impl<'a, 'tcx> Qualifier<'a, 'tcx> {
+    fn new(param_env: ty::ParameterEnvironment<'a, 'tcx>,
+           qualif_map: &'a mut DefIdMap<Qualif>,
+           mir_map: Option<&'a MirMap<'tcx>>,
+           def_id: DefId,
+           mir: &'a Mir<'tcx>,
+           mode: Mode)
+           -> Qualifier<'a, 'tcx> {
+        let mut rpo = traversal::reverse_postorder(mir);
+        let temps = promote_consts::collect_temps(mir, &mut rpo);
+        rpo.reset();
+        Qualifier {
+            mode: mode,
+            span: mir.span,
+            def_id: def_id,
+            mir: mir,
+            rpo: rpo,
+            tcx: param_env.tcx,
+            param_env: param_env,
+            qualif_map: qualif_map,
+            mir_map: mir_map,
+            temp_qualif: vec![None; mir.temp_decls.len()],
+            return_qualif: None,
+            qualif: Qualif::empty(),
+            const_fn_arg_vars: BitVector::new(mir.var_decls.len()),
+            location: Location {
+                block: START_BLOCK,
+                statement_index: 0
+            },
+            temp_promotion_state: temps,
+            promotion_candidates: vec![]
+        }
+    }
+
+    // FIXME(eddyb) we could split the errors into meaningful
+    // categories, but enabling full miri would make that
+    // slightly pointless (even with feature-gating).
+    fn unimplemented(&mut self) {
```

Comment:
can you rename this to `not_const`? I was confused reading this code for some time because I assumed that `unimplemented` would be equivalent to `bug!`, and it seemed like this would cause a fair number of panics.